modules/infra/datapath/loop_output.c

```diff
@@ -0,0 +1,53 @@
+// SPDX-License-Identifier: BSD-3-Clause
+// Copyright (c) 2024 Christophe Fontaine
+
+#include <gr.h>
+#include <gr_control_output.h>
+#include <gr_graph.h>
+#include <gr_infra.h>
+#include <gr_ip4_datapath.h>
+#include <gr_ip6_datapath.h>
+#include <gr_loopback.h>
+#include <gr_trace.h>
+
+#include <rte_graph_worker.h>
+
+enum {
+	CONTROL_OUTPUT,
+	EDGE_COUNT,
+};
+
+static uint16_t
+loop_output_process(struct rte_graph *graph, struct rte_node *node, void **objs, uint16_t nb_objs) {
+	struct control_output_mbuf_data *co;
+	struct rte_mbuf *mbuf;
+
+	for (unsigned i = 0; i < nb_objs; i++) {
+		mbuf = objs[i];
+		co = control_output_mbuf_data(mbuf);
+		co->callback = loopback_tx;
+		rte_node_enqueue_x1(graph, node, CONTROL_OUTPUT, mbuf);
+	}
+	return nb_objs;
+}
+
+static struct rte_node_register loop_output_node = {
+	.name = "loop_output",
+	.process = loop_output_process,
+	.nb_edges = EDGE_COUNT,
+	.next_nodes = {
+		[CONTROL_OUTPUT] = "control_output",
+	},
+};
+
+static void loop_output_register(void) {
+	ip_output_add_tunnel(GR_IFACE_TYPE_LOOPBACK, "loop_output");
+	ip6_output_add_tunnel(GR_IFACE_TYPE_LOOPBACK, "loop_output");
```

Comment:
I thought you would be using control_output for this. Maybe this is done in another commit. But I don't think there should be any IP level output done on loopback interfaces.

Same remark about the node name -> `s/loop_output/loopback_output/`

Reply:
`loop_output` is using `control_ouput`. Yet, you still have to format the mbuf to prepare for control output though.

Loopback interfaces are backed by a tun, which is at the L3. So, the packets must not go into eth_output.
TUN faces are similar to IPIP ifaces in that sense.

Reply:
Maybe the naming isn't right then. I was thinking of the following:

```mermaid
graph LR;
eth_input-->ip_input;
ip_input-->ip_forward;
ip_forward-->ip_output;
ip_output-->eth_output;
ip_output-->ipip_output;
ipip_output-->ip_output;
ip_input-->ip_input_local;
ip_input_local-->ipip_input;
ipip_input-->ip_input;
ip_input_local-->icmp_input;
ip_input_local-->tcp_input_local;
ip_input_local-->udp_input_local;
tcp_input_local-->loopback_input;
udp_input_local-->loopback_input;
loopback_input-->control_output;
control_input-->loopback_output;
loopback_output-->tcp_output_local;
loopback_output-->udp_output_local;
tcp_output_local-->ip_output;
udp_output_local-->ip_output;
```

Reply:
you inverted loopback_input and loopback_output. That's why you're confused !